File: 2018/day06.rb
class Day06
  def self.part1
    day06 = new(File.read('day06-input.txt').chomp)
  end

  def self.part2
  end

  attr_reader :mmx, :mmy

  def initialize(input)
    @input = input.split("\n")
  end

  def tinker
    the_world.map do |location|
      interesting_coords.map do |coord|
         { id: location == coord ? coord.id : coord.id.downcase,
           distance: location.distance(coord)
         }
      end.reject { |result| result[:id].match(/[A-Z]/)}
         .sort_by { |result| result[:distance] }
         .last(2)
    end
  end

  def interesting_coords
    @interesting_coords ||= @input.map { |line| line.split(', ') }
                                  .zip(names(@input.length))
                                  .map { |line, id| Coordinate.new(*line, id) }
  end

  def the_world

    @the_world ||= Range.new(*mmx).flat_map {|x| Range.new(*mmy).map {|y| Coordinate.new(x,y) } }
  end

  def mmx
    @mmx ||= interesting_coords.map(&:x).minmax
  end

  def mmy
    @mmy ||= interesting_coords.map(&:y).minmax
  end

  def is_edgy?(coordinate)
    mmx.include?(coordinate.x) || mmy.include?(coordinate.y)
  end

  def names(how_many)
    letters = ('A'..'Z').to_a
    numbers = (0..(how_many/letters.length)).to_a
    numbers.product(letters).map(&:join)
  end
end

class Coordinate
  attr_reader :x, :y
  attr_accessor :id

  def initialize(x, y, id='??')
    @x = x.to_i
    @y = y.to_i
    @id = id
  end

  def distance(other)
    (self.x - other.x).abs + (self.y - other.y).abs
  end

  def closest_point(points)
    points.map do |coord|
      { id: self == coord ? coord.id : coord.id.downcase,
        distance: self.distance(coord)
      }
   end.reject { |result| result[:id].match(/[A-Z]/)}
      .sort_by { |result| result[:distance] }
      # .last(2)
      # .each_cons(2)
      # .map{|p,n| p[:distance] == n[:distance] ? '..' : n[:id]}
  end

  def ==(other)
    other.is_a?(self.class) && other.x == x && other.y == y
  end
end

require 'rspec'

describe Coordinate do
  it 'can crank out coordinates' do
    coord = Coordinate.new(*'1, 9'.split(', '))
    expect(coord.x).to eq 1
    expect(coord.y).to eq 9
    expect(coord.id).to eq '??'
  end

  it 'can crank out coordinates' do
    coord = Coordinate.new(1, 9, '5G')
    expect(coord.x).to eq 1
    expect(coord.y).to eq 9
    expect(coord.id).to eq '5G'
  end

  it 'computes the manhatten distance to another coordinate' do
    a = Coordinate.new(1, 10)
    b = Coordinate.new(100, 1000)
    expect(a.distance(b)).to eq 1089
  end
end

File: 2018/day06.exs
defmodule Day06 do
  def part1 do

    finite_area_sizes(interesting_points)
  end

  def interesting_points do
    get_input |> parse_points
  end

  def get_input do
    {:ok, input} = File.read('day06-input.txt')
    input
    |> String.trim
  end

  def the_world(bounds \\ %{x: 0..400, y: 0..400}) do
    Enum.flat_map(bounds[:x], fn n -> Enum.map(bounds[:y], &{n, &1}) end)
  end

  def finite_area_sizes(points) do
    bounds = point_bounds(Map.keys(points))
    world = the_world(bounds)
    area_sizes(points, world)
    # |> Enum.filter_the_infinite_ones(ontheoutside)
    # |>
    # |> Enum.filter(fn {{_, n1}, {_, n2}} -> n1 == n2 end)
    # |> Enum.sort_by(fn {name, size} -> size end)
    # |> Enum.reverse()
    #|> hd()
  end

  def area_sizes(points, world) do
    points
    |> closest_points(world)
    # |> Map.values()
    # |> Enum.reduce(%{}, &do_sum_area_sizes/2)
  end

  defp do_sum_area_sizes(name, acc), do: Map.update(acc, name, 1, &(&1 + 1))

  def closest_points(points, world \\ the_world()) do
    world
    |> Enum.map(&{&1, closest_point(&1, points)})
    |> Enum.into(%{})
  end

  def closest_point(point, points) do
    case do_closest(point, points) |> Enum.sort() do
      [{distance, _}, {distance, _} | _] -> ".."
      [{_, other_point_name} | _] -> String.downcase(other_point_name)
    end
  end

  defp do_closest(point, points) do
    Enum.map(points, fn {other_point, other_point_name} ->
      {manhattan_distance(point, other_point), other_point_name}
    end)
  end

  def point_bounds(points) do
    {min_x, max_x} = points
                      |> Enum.map(fn {x, _} -> x end)
                      |> Enum.min_max
    {min_y, max_y} = points
                      |> Enum.map(fn {_, y} -> y end)
                      |> Enum.min_max
    %{x: Range.new(min_x, max_x),
      y: Range.new(min_y, max_y)}
  end

  def grid_locations(range \\ 0..400) do
    Enum.flat_map(range, fn n -> Enum.map(range, &{n, &1}) end)
  end

  def parse_points(input) do
    input
    |> String.split("\n")
    |> Enum.map(&parse_point/1)
    |> Enum.zip(names())
    |> Enum.into(%{})
  end

  defp parse_point(point) do
    point
    |> String.split(", ")
    |> Enum.map(&String.to_integer/1)
    |> List.to_tuple()
  end

  def names do
    ?A..?Z
    |> Enum.map(&to_string([&1]))
    |> Enum.flat_map(fn a -> Enum.map(0..1, &"#{&1}#{a}") end)
    |> Enum.sort()
  end

  def manhattan_distance({x1, y1}, {x2, y2}), do: abs(x1 - x2) + abs(y1 - y2)

end
